src/core/radio/trel_link.cpp

```diff
@@ -0,0 +1,489 @@
+        VerifyOrExit(aPacket.GetHeader().GetChannel() == mRxChannel, OT_NOOP);
+    }
+
+    if (mPanId != Mac::kPanIdBroadcast)
+    {
+        Mac::PanId rxPanId = aPacket.GetHeader().GetPanId();
+
+        VerifyOrExit((rxPanId == mPanId) || (rxPanId == Mac::kPanIdBroadcast), OT_NOOP);
+    }
+
+    // Drop packets originating from same device.
+    VerifyOrExit(aPacket.GetHeader().GetSource() != Get<Mac::Mac>().GetExtAddress(), OT_NOOP);
+
+    if (type != Header::kTypeBroadcast)
+    {
+        VerifyOrExit(aPacket.GetHeader().GetDestination() == Get<Mac::Mac>().GetExtAddress(), OT_NOOP);
+
+        if (type == Header::kTypeAck)
+        {
+            HandleAck(aPacket);
+            ExitNow();
+        }
+    }
+
+    otLogDebgMac("Trel: ReceivedPacket() [%s] plen:%d", aPacket.GetHeader().ToString().AsCString(),
+                 aPacket.GetPayloadLength());
+
+    if (aPacket.GetHeader().GetAckMode() == Header::kAckRequested)
+    {
+        SendAck(aPacket);
+    }
+
+    mRxFrame.mPsdu    = aPacket.GetPayload();
+    mRxFrame.mLength  = aPacket.GetPayloadLength();
+    mRxFrame.mChannel = aPacket.GetHeader().GetChannel();
+#if OPENTHREAD_CONFIG_MULTI_RADIO
+    mRxFrame.mRadioType = Mac::kRadioTypeTrel;
+#endif
+    mRxFrame.mInfo.mRxInfo.mTimestamp             = 0;
+    mRxFrame.mInfo.mRxInfo.mRssi                  = kRxRssi;
+    mRxFrame.mInfo.mRxInfo.mLqi                   = OT_RADIO_LQI_NONE;
+    mRxFrame.mInfo.mRxInfo.mAckedWithFramePending = true;
+
+    Get<Mac::Mac>().HandleReceivedFrame(&mRxFrame, OT_ERROR_NONE);
+
+exit:
+    return;
+}
+
+void Link::HandleAck(Packet &aAckPacket)
+{
+    otError      ackError;
+    Mac::Address srcAddress;
+    Neighbor *   neighbor;
+    uint32_t     ackNumber;
+
+    otLogDebgMac("Trel: HandleAck() [%s]", aAckPacket.GetHeader().ToString().AsCString());
+
+    srcAddress.SetExtended(aAckPacket.GetHeader().GetSource());
+    neighbor = Get<NeighborTable>().FindNeighbor(srcAddress, Neighbor::kInStateAnyExceptInvalid);
+    VerifyOrExit(neighbor != nullptr, OT_NOOP);
+
+    ackNumber = aAckPacket.GetHeader().GetPacketNumber();
+
+    // Verify that neighbor is waiting for acks and the received ack
+    // number is within the range of expected ack numbers.
+
+    VerifyOrExit(neighbor->IsRxAckNumberValid(ackNumber), OT_NOOP);
+
+    do
+    {
+        // Check whether the received ack number matches the next
+        // expected one. If it does not, it indicates that some of
+        // packets missed their acks.
+
+        ackError = (ackNumber == neighbor->GetExpectedTrelAckNumber()) ? OT_ERROR_NONE : OT_ERROR_NO_ACK;
+
+        neighbor->DecrementPendingTrelAckCount();
+
+        ReportDeferredAckStatus(*neighbor, ackError);
+        VerifyOrExit(!neighbor->IsStateInvalid(), OT_NOOP);
+
+    } while (ackError == OT_ERROR_NO_ACK);
+
+exit:
+    return;
+}
+
+void Link::SendAck(Packet &aRxPacket)
+{
+    Packet ackPacket;
+
+    ackPacket.Init(mAckPacketBuffer, sizeof(mAckPacketBuffer));
```

Comment:
Why do we define a new ACK instead of using 15.4 ACK？

Reply:
TREL ack packet serves a very different purpose/functionality compared to 15.4 ACK.
- It is used to indicate whether TREL packet was received by a known neighbor (to help update the neighbor radio link state).
- The TREL ack can be requested independent of whether the embedded/encapsulated 15.4 frame within the TREL packet requires ack or not.
    - For example, this is helpful when converting a broadcast 15.4 frame to a bunch of unicast TREL packets sent to each neighbor (the original 15.4 frame does not require ack but the TREL packets do).
    - We require  a TREL ack in response to a TREL packet destined to a known neighbor (even if the embedded 15.4 frame does not require ack).
- The TREL header include a "TREL packet number" which is used to associate a TREL packet with its TREL ack.
   - The packet number is different per neighbor and is tracked/managed by the TREL link layer.
   - The packet number is used in deferred ack mechansim which is used by TREL to ensure it can be used on WiFi links using 802.11 power-save mode.
   - 15.4 ack frame on the other hand has a rather limited format since it only contains a seq number byte value.